circom-prover/src/witness.rs

```diff
@@ -0,0 +1,63 @@
+use num::{BigInt, BigUint};
+use std::{collections::HashMap, str::FromStr};
+
+/// Witness function signature for rust_witness (inputs) -> witness
+type RustWitnessWtnsFn = fn(HashMap<String, Vec<BigInt>>) -> Vec<BigInt>;
+/// Witness function signature for witnesscalc_adapter (inputs, .dat file path) -> witness
+type WitnesscalcWtnsFn = fn(HashMap<String, Vec<BigInt>>, &str) -> Vec<BigInt>;
+
+pub enum WitnessFn {
+    WitnessCalc(WitnesscalcWtnsFn),
+    RustWitness(RustWitnessWtnsFn),
+}
+
+pub enum WitnessLib {
+    WitnessCalc,
+    RustWitness,
+}
+
+/// To create witness functions corresponding to different witness generation libs.
+#[macro_export]
+macro_rules! create_witness_fn {
+    (WitnessLib::RustWitness, $fn:ident) => {
+        rust_witness::witness!($fn);
+    };
+    (WitnessLib::WitnessCalc, $fn:ident) => {
+        witnesscalc_adapter::witness!($fn);
+    };
+    ($other:tt, $fn:ident) => {
+        compile_error!(concat!("Unknown macro type: ", stringify!($other)));
+    };
+}
+
+pub fn generate_witness(
+    witness_fn: WitnessFn,
+    inputs: HashMap<String, Vec<String>>,
+    dat_path: String,
+) -> Vec<BigUint> {
+    std::thread::spawn(move || {
```

Comment:
This thread is a noop. This `generate_witness` implementation creates a background thread, then blocks the main thread waiting for the background thread (e.g. immediately joins the thread below).

Reply:
Nice catch! I forgot to handle this. fixed in [b6d39a6](https://github.com/zkmopro/mopro/pull/304/commits/b6d39a6ba3acaddc4a0e7186b695e64ca208708a)